test/404s.test.js

```diff
@@ -96,6 +97,83 @@ test('customized 404', t => {
   })
 })
 
+test('setting a custom 404 handler multiple times is an error', t => {
+  t.plan(4)
+
+  t.test('at the root level', t => {
+    t.plan(2)
+
+    const fastify = Fastify()
+
+    fastify.setNotFoundHandler(() => {})
+    fastify.setNotFoundHandler(() => {})
+
+    fastify.listen(0, (err) => {
+      t.type(err, Error)
+      t.strictEqual(err.message, 'Not found handler already set for Fastify instance with prefix: \'/\'')
+    })
+  })
+
+  t.test('at the plugin level', t => {
+    t.plan(2)
+
+    const fastify = Fastify()
+
+    fastify.register((instance, options, next) => {
+      instance.setNotFoundHandler(() => {})
+      instance.setNotFoundHandler(() => {})
+      next()
+    }, { prefix: '/prefix' })
+
+    fastify.listen(0, (err) => {
+      t.type(err, Error)
+      t.strictEqual(err.message, 'Not found handler already set for Fastify instance with prefix: \'/prefix\'')
+    })
+  })
+
+  t.test('at the multiple levels', t => {
+    t.plan(2)
+
+    const fastify = Fastify()
+
+    fastify.register((instance, options, next) => {
+      instance.setNotFoundHandler(() => {})
+      next()
+    })
+
+    fastify.setNotFoundHandler(() => {})
+
+    fastify.listen(0, (err) => {
+      t.type(err, Error)
+      t.strictEqual(err.message, 'Not found handler already set for Fastify instance with prefix: \'/\'')
+    })
+  })
```

Comment:
I’m -1 to the behavior showed by this specific test. This should be allowed: if a deeper level sets a notFound handler, that should be called and not the one of the parent.

Reply:
I understand the confusion here. This is a complicated part of Fastify. I can explain why this error is helpful though.

The reason is that notFound handlers cannot be fully encapsulated. They can only be encapsulated by plugins with a `prefix`. To compare, error handlers can be fully encapsulated because they get attached to the `context` of routes. However, notFound handlers are their own routes on a different `find-my-way` instance (with wildcard URLs such as `/*`, `/prefix/*`, `/another/prefix/*`, etc.).

This means that these two calls:

1.
```js
fastify.setNotFoundHandler(function handlerA (request, reply) { })

fastify.setNotFoundHandler(function handlerB (request, reply) { })
```
2.
```js
fastify.setNotFoundHandler(function handlerA (request, reply) { })

fastify.register((instance, options, next) => {
  instance.setNotFoundHandler(function handlerB (request, reply) { })
  next()
})
```

are exactly the same thing. Before this change, `handlerB` would [silently replace](https://github.com/fastify/fastify/blob/c3a80474525ed8e827acb5d44364c514f5475320/fastify.js#L659) `handlerA` so `handlerA` would be completely lost (then all not-found requests would be handled by `handlerB`).

https://github.com/fastify/fastify/blob/c3a80474525ed8e827acb5d44364c514f5475320/fastify.js#L659

The error is helpful because it prevents people from overwriting their own notFound handlers ("shooting themselves in the foot").

Reply:
👍  you are right.

Reply:
:) I'll update the docs to help explain this to users.